fast_llm/layers/transformer/config.py

```diff
@@ -114,48 +132,153 @@ class RotaryConfig(BaseModelConfig):
+        angles = torch.outer(positions, self._get_angle_scales(kv_channels, device))
+        frequencies = torch.polar(torch.ones_like(angles), angles)[None, :, None, :].to(torch.complex64)
+        if not self.complex_format:
+            frequencies = convert_rotary_complex_to_real(
+                torch.view_as_real(frequencies).flatten(-2), kv_channels, 3
+            ).contiguous()
+        return frequencies
+
+    def _get_angle_scales(self, kv_channels: int, device="cuda") -> "torch.Tensor":
+        import torch
+
+        return self.theta ** -torch.arange(0, 1, 2 / kv_channels, device=device, dtype=torch.float64)
+
+
+@config_class()
+class Llama3RotaryConfig(DefaultRotaryConfig):
+    """
+    Llama3 scaling: https://github.com/meta-llama/llama-models/blob/baf7b01b6e62bc7126c7b558d2b67d4533142680/models/llama3/reference_impl/model.py#L45-L67
+    """
+
+    # TODO: Add descriptions.
+    scale_factor: float = Field(default=8.0, hint=FieldHint.feature)
+    low_frequency_factor: float = Field(default=1.0, hint=FieldHint.feature)
+    high_frequency_factor: float = Field(default=4.0, hint=FieldHint.feature)
+    original_context_length: int = Field(default=8192, hint=FieldHint.feature)
+
+    def _validate(self) -> None:
+        super()._validate()
+        Assert.gt(self.high_frequency_factor, self.low_frequency_factor)
+
+    def _get_angle_scales(self, kv_channels: int, device="cuda") -> "torch.Tensor":
+        import torch
+
+        scales = super()._get_angle_scales(kv_channels, device)
+        low_frequency_wavelength = self.original_context_length / self.low_frequency_factor
+        high_frequency_wavelength = self.original_context_length / self.high_frequency_factor
+        new_scales = []
+        for scale in scales:
+            wavelength = 2 * math.pi / scale
+            if wavelength < high_frequency_wavelength:
+                new_scales.append(scale)
+            elif wavelength > low_frequency_wavelength:
+                new_scales.append(scale / self.scale_factor)
+            else:
+                smooth = (self.original_context_length / wavelength - self.low_frequency_factor) / (
+                    self.high_frequency_factor - self.low_frequency_factor
+                )
+                new_scales.append((1 - smooth) * scale / self.scale_factor + smooth * scale)
+        return torch.stack(new_scales)
+
+
+@config_class()
+class YarnRotaryConfig(DefaultRotaryConfig):
+    """
+    Yarn scaling:
+    https://github.com/huggingface/transformers/blob/006d9249ec0270ff6c4d3840979d23fe94bdc763/src/transformers/modeling_rope_utils.py#L163
+    [original paper](https://arxiv.org/abs/2309.00071)
+    """
+
+    # TODO: Add descriptions.
+    scale_factor: float = Field(default=8.0, hint=FieldHint.feature)
     attention_factor: None | float = Field(
         default=None,
-        desc="Attention factor for yarn-type scaling.",
```

Comment:
is this because the desc is not informative or because something has changed so that descriptions don't work anymore?

Reply:
`Attention factor` is just repeating the field name, and `for yarn-type scaling` is now redundant with the type/class name. So the existing descriptions are completely useless.

Reply:
ok, redundancy, I get it. But rather than dropping the description entirely, should we consider replacing it with something more informative? e.g., when this field matters, what values are valid, what happens if it's None, etc. Otherwise we're losing a chance to document intent.

Reply:
These descriptions were already irrelevant, so whether I'm removing them or leaving them as a repetition of the field name doesn't matter. And more importantly, I'm not familiar enough with these classes to make a proper description, feel free to suggest some.